examples/hal/quad/main.rs

```diff
@@ -4,613 +4,1357 @@
 )]
 
 extern crate env_logger;
-extern crate gfx_hal as hal;
 #[cfg(feature = "dx12")]
 extern crate gfx_backend_dx12 as back;
-#[cfg(feature = "vulkan")]
-extern crate gfx_backend_vulkan as back;
-#[cfg(feature = "metal")]
-extern crate gfx_backend_metal as back;
 #[cfg(feature = "gl")]
 extern crate gfx_backend_gl as back;
-
-extern crate winit;
+#[cfg(feature = "metal")]
+extern crate gfx_backend_metal as back;
+#[cfg(feature = "vulkan")]
+extern crate gfx_backend_vulkan as back;
+extern crate gfx_hal as hal;
 extern crate image;
+extern crate takeable_option;
+extern crate winit;
 
-use hal::{buffer, command, format as f, image as i, memory as m, pass, pso, pool};
-use hal::{Device, Instance, PhysicalDevice, Surface, Swapchain};
-use hal::{
-    DescriptorPool, FrameSync, Primitive,
-    Backbuffer, SwapchainConfig,
-};
-use hal::format::{AsFormat, ChannelType, Rgba8Srgb as ColorFormat, Swizzle};
-use hal::pass::Subpass;
-use hal::pso::{PipelineStage, ShaderStageFlags, Specialization};
-use hal::queue::Submission;
-
-use std::io::Cursor;
-
-const ENTRY_NAME: &str = "main";
-
-#[derive(Debug, Clone, Copy)]
-#[allow(non_snake_case)]
-struct Vertex {
-    a_Pos: [f32; 2],
-    a_Uv: [f32; 2],
-}
+#[macro_use]
+extern crate glsl_to_spirv_macros;
```

Comment:
why do we need those macros? I'd prefer us to just do the stuff at run time if it makes us depend on less externals